### PR TITLE
#160452487 - Updates article endpoints to support tagging

### PR DESCRIPTION
**Pull request by macphilips**
#### What does this PR do?
- It adds support for tagging an article.
#### Description of Task to be completed?
- Updated test suites for article endpoints.
- Implements tagging feature that lets authors an article on the article endpoints.
- Updated the `UserProfile` and `UserController` files.
- Renamed `HttpError.throw404ErrorIfNull` to `HttpError.throwErrorIfNull`
 to make it reusagable for other kind of errors. 
- Rename the `procfile` to `Procfile` to enable Heroku run migration upon deployment.
- Updated the other test suites to increase coverage.
- Refactored `ArticleValidator` and `AuthValidator` file
#### How should this be manually tested?
- Fetch and checkout to the branch
```$xslt
> git fetch
> git checkout ft-160452487-tag-authors-article
> npm install 
```
- Setup database using .env-sample
* Run Sequelize migration 
```$xslt
 > npm run migration
 ```
- Run application
```$xslt 
> npm run start
```
- To run test
```$xslt 
> npm run test
```

* Using a REST client such as postman to test endpoints: 
To create an article: `POST /api/v1/articles` 
To get list of articles: `GET /api/v1/articles`,
To get, update, or delete a specific article:
`GET /api/v1/articles/:slug`.
`PUT /api/v1/articles/:slug`.
`DELETE /api/v1/articles/:slug`.

sample request body
```json
{
     "title": "my awesome article",
     "description": "This is the best description I can think of right now",
     "body": "Nunc sed diam suscipit, lobortis eros nec, auctor nisl. Nunc ac magna non justo varius rutrum sit amet feugiat elit. ",
     "tags": [ "Authors Haven" ]
}
```
* For API Documuntation
http://localhost:3000/api-docs/

#### Any background context you want to provide?
- None
#### What are the relevant pivotal tracker stories?
- [#160452487](<https://www.pivotaltracker.com/story/show/160452487>)
#### Questions:

### CODE REVIEW

File: helpers/Util.js [hunk 1]
```diff
@@ -74,4 +74,21 @@ export default class Util {
...
   }
+
+  /**
+   *
+   * @param {Array} fieldNames
+   * @param {object} data
+   * @return {object} Gets the fields in the array from the data object if they exist,
+   * and return those field as an object
+   */
+  static extractFieldsFromObject(fieldNames, data) {
```
Comment (olusoladavid) [inline helpers/Util.js:87]: Great job on the refactoring here.

File: models/index.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ import config from '../config/config';
 
 const basename = path.basename(__filename);
 const env = process.env.NODE_ENV || 'development';
-const isDev = env === 'development';
+const isDev = env === 'developments';
```
Comment (olusoladavid) [inline models/index.js]: Why did you change `development` to `developments` here?

File: helpers/TagHelper.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+   *
+   * @param {Array}tags
+   * @return {Array} Returns array of tag object to be stored in the databse
+
+   */
+  static getTagObj(tags) {
+    const result = [];
+    if (tags) {
+      tags.forEach((tag) => {
+        result.push(Tag.findOrCreate({ where: { name: tag } }));
```
Comment (olusoladavid) [inline helpers/TagHelper.js:34]: Nice work on making the database operations fully asynchronous. I think it will be nice to check if the tag is an empty string too. The current implementation allows empty strings to be stored to the database as tags.

Comment (macphilips) [reply helpers/TagHelper.js:34]: Nice catch. Thanks

Comment (olusoladavid) [reply helpers/TagHelper.js:34]: Have you resolved this?
